### ISSUE TITLE
how to do customization in Edirom (CSS and Javascript)

### ISSUE THREAD

**Issue by peterstadler** [Edirom/Edirom-Online#648]
IFF I understand correctly, at present only CSS (not JS) injection is possible via eXist's REST interface.

My proposal is to create an Edirom way to inject both JS and CSS, not relying on the REST interface since this should IMHO be disabled for production use. It'll need 

* a user preference setting (per edition) that will depict the path to a CSS and a JS folder in the data package
* a function to insert every CSS (and JS) file found within those folders into the Edirom HTML
* a Edirom controller rule to forward everything under e.g. `custom-resources/$edition-id$/css` to the respective collection in the data package

If this seems a viable approach I'm happy to work on this and send in a PR!

**Comment by daniel-jettka** [comment]
ping @bwbohl @peterstadler What do you think about this one at the moment?

**Comment by daniel-jettka** [comment]
Proposal from community meeting:

- define fixed paths for css and js directories in data package
- make backend read the css and js from data package and populate to frontend
- frontend applies custom css and js
- test case: Bargheer edition styles/scripts https://github.com/Edirom/Bargheer-Edition/issues/13

**Comment by feuerbart** [comment]
OPERA used lots of custom css, js and xquery, e.g.
`<entry key="plugin_textview" value="../../../exist/rest/db/contents/edition-74338567/js/TextViewPlugin.js"/>`
`<entry key="data/xql/getLinkTarget.xql" value="../../../exist/rest/db/contents/edition-74338567/xql/getLinkTarget.xql"/>`

OPEN Faust uses [`additional_css_path`](https://git.uni-paderborn.de/kio/open-faust/-/blob/main/content/prefs/prefs.xml?ref_type=heads#L34) for css and only xquery for the text edition:
[`<entry key="data/xql/getText.xql" value="apps/edirom/open-faust/content/xql/getText.xql"/>`](https://git.uni-paderborn.de/kio/open-faust/-/blob/main/content/prefs/prefs.xml?ref_type=heads#L40)

TaMuDi only uses [`additional_css_path`](https://git.uni-paderborn.de/tanzmusikdigital/edition/edition-data/-/blob/main/content/prefs.xml?ref_type=heads#L32)

**Comment by daniel-jettka** [comment]
@bwbohl @feuerbart So you say it's working (and just needs documentation - or does it exist already?!), right?

**Comment by bwbohl** [comment]
> OPERA used lots of custom css, js and xquery, e.g. `<entry key="plugin_textview" value="../../../exist/rest/db/contents/edition-74338567/js/TextViewPlugin.js"/>` `<entry key="data/xql/getLinkTarget.xql" value="../../../exist/rest/db/contents/edition-74338567/xql/getLinkTarget.xql"/>`
> 
> OPEN Faust uses […] xquery for the text edition: [`<entry key="data/xql/getText.xql" value="apps/edirom/open-faust/content/xql/getText.xql"/>`](https://git.uni-paderborn.de/kio/open-faust/-/blob/main/content/prefs/prefs.xml?ref_type=heads#L40)
> […]

@feuerbart do the custom XQueries get applied when you run eXist.-db in production mode? That would be surprising

**Comment by bwbohl** [comment]
> [@bwbohl](https://github.com/bwbohl) [@feuerbart](https://github.com/feuerbart) So you say it’s working (and just needs documentation - or does it exist already?!), right?

Depends on what “it” is:
- CSS yes by calculating the frontend path and adding it (cf. Edirom/Edirom-Online-Frontend#78)
- JS no, especially not in the way it used to work
- (Backend: XQuery (for completeness) should not work when access to eXist-db REST interface is deactivated, that’s why we went  for another solution with the CSS stuff)

The good thing about the CSS approach is that it could be changed to work with any URI. The approach could probably also be used for the other two. Nevertheless, I don’t think it’s a good idea to go down that path and rely on potentially unavailable external resources, especially the JS (and XQuery) plugin mechanism used to override whole files or individual functions. The soundest approach, from my perspective, is to define a fixed location within the edition data package. This is also to minimise attack vectors, such as REST endpoints. Moreover, the JS mechanism probably should not use `eval()` as the `PreferenceController` does in the current code.

With predefined locations, the prefs would not have to provide the individual locations. We could probably introduce pref entries to activate or deactivate overrides (for XQuery, JS, CSS, etc.), which might come in handy for debugging.

**TLDR: The application should inspect fixed locations within the edition package. No paths in prefs.xml, no REST parameters.**

So here would be a first approach for a defined structure in an edition’s XAR:

```txt
10:11edition-ID/
├── overrides/
│   ├── css/
│   │   └── *.css                 # Styles that modify/extend EO
│   ├── js/
│   │   └── *.js                  # Code that uses EO's override pattern
│   ├── xql/
│   │   └── *.xql                 # Whole-file replacement
│   ├── xqm/
│   │   └── *.xqm                 # Module function replacement via function lookup
│   └── xsl/
│       └── *.xsl                 # Template-level override via import
├── resources/              # additional libraries or assets
│   ├── css/
│   │   └── *.css                 # Third-party library styles
│   ├── js/
│   │   └── *.js                  # Third-party libraries
│   ├── images/
│   └── fonts/
├── locale/
│   ├── lang_de.xml               # Existing, prefs-configured with fallback / could also be changed to default lookup
│   └── lang_en.xml
├── data/                         # The edition data
├── prefs.xml                     # edition specific edirom prefs
└── edition-ID.xml                # Edition descriptor
```

**Comment by krHERO** [comment]
we talked about this topic at the [community meeting 4.3.2026](https://github.com/Edirom/Edirom-Online/wiki/2026-03-04-Edirom-Community-Meeting) and decided to go on with the discussion in this Edirom Online main repo

@daniel-jettka volunteered to write some documentation about the current working status, thanks! https://github.com/Edirom/Edirom-Online/pull/651

**Comment by daniel-jettka** [comment]
assigned myself to document the js injection via prefs.xml -> entry "plugin_..."
(see e.g. Bargheer Edition)
